numpy/ma/core.py

```diff
@@ -3121,6 +3121,13 @@ def __getitem__(self, indx):
             if isinstance(indx, basestring):
                 if self._fill_value is not None:
                     dout._fill_value = self._fill_value[indx]
+                    # If we're indexing a multidimensional field in a
+                    # structured array (such as dtype("(2)i2,(2)i1")),
+                    # indexing fill_value will yield a multi-element
```

Comment:
Short of teaching numpy to give us a way to create such arrays, my guess is, it is the only solution. And I am not sure you can easily stop numpy. I expect it is a too large change even if for the indexing result of masked arrays themself.

However, I think I would prefer a UserWarning or even an error when it happens that the individual fill_values differe (though to point out, NaN is equivalence of course). Normally it will not be the case and all is fine, and if it happens someone maybe wants to know about it.